src/devtools/pachdev/actions.go

```diff
@@ -0,0 +1,571 @@
+			images[i] = images[j]
+		}
+		i++
+	}
+	if i < len(images) {
+		return images[:i], nil
+	}
+	return images, nil
+}
+
+func (op *deployOp) getOldPachVersion() (string, error) {
+	// Call the 'Version' API to get the Pachyderm version
+	resp, err := op.PachClient().GetVersion(op.Ctx(), &emptypb.Empty{})
+	if err != nil {
+		return "", err
+	}
+
+	return version.PrettyPrintVersion(resp), nil
+}
+
+func (op *deployOp) getPachdPodName() (string, error) {
+	// Retrieve the list of pachd pods (matching the below label selector)
+	labelSelector := labels.Set{
+		"suite": "pachyderm",
+		"app":   "pachd",
+	}.AsSelector()
+	podList, err := op.KubeClient().CoreV1().Pods("default").List(context.TODO(), metav1.ListOptions{
+		LabelSelector: labelSelector.String(),
+	})
+	if err != nil {
+		return "", err
+	}
+
+	// Check if there is a running pachd pod
+	if len(podList.Items) == 0 {
+		return "", fmt.Errorf("no running pachd pod found")
+	}
+
+	// Get the name of the first pachd pod
+	pachdPodName := podList.Items[0].Name
+
+	return pachdPodName, nil
+}
+
+func (op *deployOp) maybeBuildOrPullPachyderm() {
+	// Not implemented
+}
+
+// maybeDeleteK8sCluster deletes the kind cluster if it exists.
+// Taken from: https://github.com/kubernetes-sigs/kind/blob/8f6cb8f45de2c56e4d13de1b4678f1d538755208/pkg/cmd/kind/delete/cluster/deletecluster.go#L68-L77
+func (op *deployOp) maybeDeleteK8sCluster() error {
+	if err := op.KindProvider().Delete(KindTestClusterName, ""); err != nil {
+		op.err = errors.Wrap(err, "failed to delete kind cluster")
+		return op.err
+	}
+	return nil
+}
+
+// maybeCreateK8sCluster creates the kind cluster if it doesn't exist.
+func (op *deployOp) maybeCreateK8sCluster() error {
+	// TODO(msteffen) maybe switch to sigs.k8s.io/kind/pkg/cluster?
+
+	// Check if the cluster already exists
+	// The implementation of `kind get clusters` is here:
+	// https://github.com/kubernetes-sigs/kind/blob/8f6cb8f45de2c56e4d13de1b4678f1d538755208/pkg/cmd/kind/get/clusters/clusters.go#L48-L63
+	clusters, err := op.KindProvider().List()
+	if err != nil {
+		op.err = errors.Wrap(err, "could not list existing kind clusters")
+		return op.err
+	}
+	for _, cluster := range clusters {
+		if strings.Contains(cluster, KindTestClusterName) {
+			return nil
+		}
+	}
+
+	// Create the cluster
+	// The implementation of `kind create cluster` is here:
+	// https://github.com/kubernetes-sigs/kind/blob/8f6cb8f45de2c56e4d13de1b4678f1d538755208/pkg/cmd/kind/create/cluster/createcluster.go#L98-L123
+	if err = op.KindProvider().Create(
+		KindTestClusterName,
+		cluster.CreateWithDisplayUsage(true),
+		cluster.CreateWithDisplaySalutation(true),
+	); err != nil {
+		op.err = errors.Wrap(err, "failed to create cluster")
+		return op.err
+	}
+	return nil
+}
+
+func (op *deployOp) maybeDeployPachyderm() {
+	op.
+}
+
+func (op *deployOp) Deploy() {
+	op.maybeDeleteK8sCluster()
+
+	eg := &errgroup.Group{}
+	eg.Go(op.maybeCreateK8sCluster)
+	// eg.Go(op.maybeBuildOrPullPachyderm)
```

Comment:
I think it's fine to require that the build be done before we create the k8s cluster.

The way I start things off in my setup.sh script is by pulling the latest version (just hard-coded when I remember to update it), and then waiting until that's running.  The rebuild.sh script builds pachyderm first, then does all the docker and k8s things that need to be done.  

I wouldn't invest too much time shelling out to Docker to do the builds, since I hope that will go away soon... just a stub shell script that does "go build whatever" "docker build -f Dockerfile.pachd" "docker tag" that then calls this will probably be fine.  When we bazel-ify this script, the OCI images will just be a dependency of the "update my pachyderm" target, so that will be when we have a "no shell script" version.  But for a new engineer starting on 1/2, maybe they'll be OK with a tiny tiny bash script for a couple weeks ;)